scd/token.c

```diff
@@ -0,0 +1,321 @@
+	    unsigned char **wrapped_key, int *wrapped_key_len)
+{
+	return softtoken_wrap_key(token->token_data->int_token.softtoken, plain_key, plain_key_len,
+				  wrapped_key, wrapped_key_len);
+}
+
+int
+int_unwrap_st(scd_token_t *token, UNUSED char *label, unsigned char *wrapped_key,
+	      size_t wrapped_key_len, unsigned char **plain_key, int *plain_key_len)
+{
+	return softtoken_unwrap_key(token->token_data->int_token.softtoken, wrapped_key,
+				    wrapped_key_len, plain_key, plain_key_len);
+}
+
+/* TODO: add token provisioning for softtoken */
+int
+int_change_pw_st(scd_token_t *token, const char *oldpass, const char *newpass,
+		 UNUSED unsigned char *pairing_secret, UNUSED size_t pairing_sec_len,
+		 UNUSED bool is_provisioning)
+{
+	return softtoken_change_passphrase(token->token_data->int_token.softtoken, oldpass,
+					   newpass);
+}
+
+/*  -----------------------------------------------------------------------  */
+#ifdef ENABLESCHSM
+
+int
+int_lock_usb(scd_token_t *token)
+{
+	return usbtoken_lock(token->token_data->int_token.usbtoken);
+}
+
+int
+int_unlock_usb(scd_token_t *token, char *passwd, unsigned char *pairing_secret,
+	       size_t pairing_sec_len)
+{
+	TRACE("SCD: int_usb_unlock");
+	return usbtoken_unlock(token->token_data->int_token.usbtoken, passwd, pairing_secret,
+			       pairing_sec_len);
+}
+
+bool
+int_is_locked_usb(scd_token_t *token)
+{
+	return usbtoken_is_locked(token->token_data->int_token.usbtoken);
+}
+
+bool
+int_is_locked_till_reboot_usb(scd_token_t *token)
+{
+	return usbtoken_is_locked_till_reboot(token->token_data->int_token.usbtoken);
+}
+
+int
+int_wrap_usb(scd_token_t *token, char *label, unsigned char *plain_key, size_t plain_key_len,
+	     unsigned char **wrapped_key, int *wrapped_key_len)
+{
+	return usbtoken_wrap_key(token->token_data->int_token.usbtoken, (unsigned char *)label,
+				 strlen(label), plain_key, plain_key_len, wrapped_key,
+				 wrapped_key_len);
+}
+
+int
+int_unwrap_usb(scd_token_t *token, char *label, unsigned char *wrapped_key, size_t wrapped_key_len,
+	       unsigned char **plain_key, int *plain_key_len)
+{
+	return usbtoken_unwrap_key(token->token_data->int_token.usbtoken, (unsigned char *)label,
+				   strlen(label), wrapped_key, wrapped_key_len, plain_key,
+				   plain_key_len);
+}
+
+int
+int_change_pw_usb(scd_token_t *token, const char *oldpass, const char *newpass,
+		  unsigned char *pairing_secret, size_t pairing_sec_len, bool is_provisioning)
+{
+	return usbtoken_change_passphrase(token->token_data->int_token.usbtoken, oldpass, newpass,
+					  pairing_secret, pairing_sec_len, is_provisioning);
+}
+#endif // ENABLESCHSM
+
+/*  -----------------------------------------------------------------------  */
+
+/**
+ * creates a new generic token
+ * calls the respective create function for the selected type of token and
+ * sets the function pointer appropriately
+ */
+scd_token_t *
+token_new(token_constr_data_t *constr_data)
```

Comment:
const token_constr_data_t const *

Reply:
i believe the second const is a) not necessary as the pointer itself is just a local copy b) if you disagree, shouldn't it be after the '*'?